File: web/src/lib/fs-utils.ts
import path from 'path'
import os from 'os'
import fs from 'fs'

/**
 * Returns a writable temporary directory path and ensures it exists.
 * In production (Vercel), we prioritize /tmp.
 * In development, we use .tmp in the execution directory.
 */
export function getSafeTmpDir(subDir: string = ''): string {
    // AGGRESSIVE PRODUCTION FALLBACK: Try system /tmp first, then os.tmpdir, then local.
    // This removes reliance on environment variables that might be missing in certain contexts.
    let baseDir = '/tmp'
    
    try {
        if (!fs.existsSync(baseDir)) {
            baseDir = os.tmpdir()
        }
    } catch (e) {
        baseDir = path.resolve(process.cwd(), '../execution/.tmp')
    }

    const targetDir = subDir ? path.join(baseDir, subDir) : baseDir

    try {
        if (!fs.existsSync(targetDir)) {
            fs.mkdirSync(targetDir, { recursive: true })
        }
    } catch (e) {
        // Final fallback if /tmp is strictly forbidden (unlikely on Vercel)
        const fallback = path.resolve(process.cwd(), '../execution/.tmp', subDir)
        if (!fs.existsSync(fallback)) {
            fs.mkdirSync(fallback, { recursive: true })
        }
        return fallback
    }

    return targetDir
}

export function getSafeTmpPath(fileName: string, subDir: string = ''): string {
    return path.join(getSafeTmpDir(subDir), fileName)
}

File: execution/fs_utils.py
import os
import tempfile
import sys

def get_safe_tmp_dir(subdir: str = None) -> str:
    """
    Returns a writable directory path.
    Prioritizes system /tmp for production environments (Vercel/Railway).
    """
    # Use environment variable if set, otherwise fallback to system tmp
    base_dir = os.environ.get("DISTILL_TMP_DIR")
    
    if not base_dir:
        if os.path.exists('/tmp'):
            base_dir = '/tmp'
        else:
            # Local fallback for extreme cases using standard library
            base_dir = tempfile.gettempdir()

    if subdir:
        target = os.path.join(base_dir, subdir)
        os.makedirs(target, exist_ok=True)
        return target
        
    os.makedirs(base_dir, exist_ok=True)
    return base_dir

def get_safe_tmp_path(filename: str, subdir: str = None) -> str:
    """
    Returns a full path to a file in the safe tmp directory.
    """
    return os.path.join(get_safe_tmp_dir(subdir), filename)
